File: src/core/analyzer.rs
use apica_common::bytecodes::ApicaEntrypointBytecode;
use crate::nodes::compound::NodeCompound;
use crate::nodes::global_scope::NodeGlobalScope;
use crate::nodes::node::Node;
use crate::utils::diagnostic::{Diagnostic, DiagnosticKind};
use crate::utils::diagnostic_bag::DiagnosticBag;

pub struct Analyzer<'a> {
    diag_bag: &'a mut DiagnosticBag,
}

impl<'a> Analyzer<'a> {
    pub fn init(diag_bag: &'a mut DiagnosticBag) -> Analyzer<'a> {
        Analyzer { diag_bag }
    }

    pub fn analyze(&mut self, root: &mut NodeCompound) {
        if !self.check_entrypoint_specs(root) {
            return;
        }

        self.extract_specs(root);
        self.extract_globals(root);
    }

    fn check_entrypoint_specs(&mut self, root: &mut NodeCompound) -> bool {
        let mut entry_init: u8 = 0;
        let mut entry_update: u8 = 0;
        let mut entry_quit: u8 = 0;
        let mut specs: u8 = 0;
        
        let mut success: bool = true;

        for node in root.get_nodes() {
            if let Node::Entrypoint(entry) = node {
                match entry.get_entry() {
                    ApicaEntrypointBytecode::Init => entry_init = if entry_init == 0 { 1 } else { 2 },
                    ApicaEntrypointBytecode::Update => entry_update = if entry_update == 0 { 1 } else { 2 },
                    ApicaEntrypointBytecode::Quit => entry_quit = if entry_quit == 0 { 1 } else { 2 },
                }
            } else if let Node::DataSpecs(_) = node {
                specs = if specs == 0 { 1 } else { 2 };
            }
        }

        if entry_init != 1 {
            self.diag_bag.add(Diagnostic::init_message(
                DiagnosticKind::Error,
                String::from("AnalyzerError: A correct main apica source file should contain a unique `init` entrypoint"),
            ));
            success = false;
        }

        if entry_update != 1 {
            self.diag_bag.add(Diagnostic::init_message(
                DiagnosticKind::Error,
                String::from("AnalyzerError: A correct main apica source file should contain a unique `update` entrypoint"),
            ));
            success = false;
        }

        if entry_quit != 1 {
            self.diag_bag.add(Diagnostic::init_message(
                DiagnosticKind::Error,
                String::from("AnalyzerError: A correct main apica source file should contain a unique `quit` entrypoint"),
            ));
            success = false;
        }
        
        if specs != 1 {
            self.diag_bag.add(Diagnostic::init_message(
                DiagnosticKind::Error,
                String::from("AnalyzerError: A correct main apica source file should contain a unique `specs` declaration")
            ));
            success = false;
        }

        success
    }

    fn extract_globals(&mut self, root: &mut NodeCompound) {
        let mut extracted = vec![];
        let nodes = root.get_mut_nodes();
        let mut kept = vec![];

        for node in nodes.drain(..) {
            if matches!(node, Node::Entrypoint(_)) || matches!(node, Node::DataSpecs(_)) || matches!(node, Node::EndOfFile(_)) {
                kept.push(node);
            } else {
                extracted.push(node);
            }
        }

        *nodes = kept;
        if !extracted.is_empty() {
            for node in root.get_mut_nodes() {
                if let Node::Entrypoint(entry) = node {
                    if *entry.get_entry() == ApicaEntrypointBytecode::Init {
                        if let Node::Compound(entry_compound) = entry.get_mut_body() {
                            let init_nodes = entry_compound.get_mut_nodes();

                            init_nodes.insert(0, Node::GlobalScope(Box::new(
                                NodeGlobalScope::init(
                                    extracted[0].get_position().clone(),
                                    Node::Compound(NodeCompound::init(
                                        extracted[0].get_position().clone(),
                                        extracted,
                                    )),
                                ),
                            )));

                            break;
                        }
                    }
                }
            }
        }

    }

    fn extract_specs(&mut self, root: &mut NodeCompound) {
        let nodes = root.get_mut_nodes();
        let mut specs = None;

        let mut i = 0;
        while i < nodes.len() {
            if matches!(nodes[i], Node::Entrypoint(_) | Node::EndOfFile(_)) {
                i += 1;
            } else {
                specs = Some(nodes.remove(i));
            }
        }

        if let Some(spec) = specs {
            nodes.insert(0, spec);
        }
    }
}

File: src/core/analyzer.cpp
#include "core/analyzer.hpp"
#include "nodes/entrypoint.hpp"
#include "utils/diagnostic_bag.hpp"
#include "utils/errors.hpp"
#include "nodes/data/specs.hpp"
#include "nodes/global_scope.hpp"
#include "utils/id_generator.hpp"

using namespace core;

Analyzer::Analyzer() {

}

void Analyzer::analyze(nodes::NodeCompound *root, bool cmp) {
    if (cmp && !this->checkEntrypoints(root))
        return;
    
    this->setIds(root);
    this->extractSpecifications(root);
    this->extractGlobals(root);
}

bool Analyzer::checkEntrypoints(nodes::NodeCompound *root) {
    uint8_t entry_init = 0, entry_update = 0, entry_quit = 0;
    bool success = true;

    for (nodes::Node *node : root->getNodes()) {
        if (node->getKind() == nodes::NodeKind::Entrypoint) {
            nodes::NodeEntrypoint *entrypoint = static_cast<nodes::NodeEntrypoint*>(node);
            switch (entrypoint->getEntryBytecode()) {
                case common::bytecodes::ApicaEntrypointBytecode::Init:
                    entry_init = entry_init == 0 ? 1 : 2;
                    break;
                
                case common::bytecodes::ApicaEntrypointBytecode::Update:
                    entry_update = entry_update == 0 ? 1 : 2;
                    break;
                
                case common::bytecodes::ApicaEntrypointBytecode::Quit:
                    entry_quit = entry_quit == 0 ? 1 : 2;
                    break;

                default: break;
            }
        }
    }

    if (entry_init != 1) {
        utils::DiagnosticBag::getInstance().addDiagnostic(utils::Diagnostic(
            utils::DiagnosticKind::Error,
            std::string(utils::ANL_ERROR_NO_UNIQUE_INIT)
        ));
        success = false;
    }

    if (entry_update != 1) {
        utils::DiagnosticBag::getInstance().addDiagnostic(utils::Diagnostic(
            utils::DiagnosticKind::Error,
            std::string(utils::ANL_ERROR_NO_UNIQUE_UPDATE)
        ));
        success = false;
    }

    if (entry_quit != 1) {
        utils::DiagnosticBag::getInstance().addDiagnostic(utils::Diagnostic(
            utils::DiagnosticKind::Error,
            std::string(utils::ANL_ERROR_NO_UNIQUE_QUIT)
        ));
        success = false;
    }

    return success;
}

bool Analyzer::checkSpecifications(nodes::NodeCompound *root) {
    uint8_t specs = 0;
    bool spec_title = false, spec_id = false, spec_version = false;

    bool success = true;

    std::optional<nodes::NodeDataSpecifications*> node_specs = std::nullopt;
    for (nodes::Node *node : root->getNodes()) {
        if (node->getKind() == nodes::NodeKind::DataSpecs) {
            specs = specs == 0 ? 1 : 2;
            node_specs = static_cast<nodes::NodeDataSpecifications*>(node);
        }
    }

    if (specs != 1) {
        utils::DiagnosticBag::getInstance().addDiagnostic(utils::Diagnostic(
            utils::DiagnosticKind::Error,
            std::string(utils::ANL_ERROR_NO_SPECS)
        ));
        return false;
    }

    for (nodes::NodeData *data : node_specs.value()->getDataNodes()) {
        switch (data->getSpecificationBytecode()) {
            case common::bytecodes::ApicaSpecificationBytecode::Title:
                spec_title = true;
                break;
            
            case common::bytecodes::ApicaSpecificationBytecode::Id:
                spec_id = true;
                break;
            
            case common::bytecodes::ApicaSpecificationBytecode::Version:
                spec_version = true;
                break;

            default: break;
        }
    }

    if (!spec_title) {
        utils::DiagnosticBag::getInstance().addDiagnostic(utils::Diagnostic(
            utils::DiagnosticKind::Error,
            std::string(utils::ANL_ERROR_NO_TITLE_SPEC)
        ));
        success = false;
    }

    if (!spec_id) {
        utils::DiagnosticBag::getInstance().addDiagnostic(utils::Diagnostic(
            utils::DiagnosticKind::Error,
            std::string(utils::ANL_ERROR_NO_ID_SPEC)
        ));
        success = false;
    }

    if (!spec_version) {
        utils::DiagnosticBag::getInstance().addDiagnostic(utils::Diagnostic(
            utils::DiagnosticKind::Error,
            std::string(utils::ANL_ERROR_NO_VERSION_SPEC)
        ));
        success = false;
    }

    return success;
}

void Analyzer::extractGlobals(nodes::NodeCompound *root) {
    std::vector<nodes::Node*> extracted;
    std::vector<nodes::Node*> &mutable_nodes = root->getMutableNodes();

    for (auto it = mutable_nodes.begin(); it != mutable_nodes.end(); ) {
        nodes::Node* node = *it;

        if (node->getKind() == nodes::NodeKind::Entrypoint ||
            node->getKind() == nodes::NodeKind::DataSpecs ||
            node->getKind() == nodes::NodeKind::EndOfFile) {
            ++it;
        } else {
            extracted.push_back(node);
            it = mutable_nodes.erase(it);
        }
    }

    if (!extracted.empty()) {
        std::optional<nodes::NodeEntrypoint*> entrypoint_init = std::nullopt;
        for (nodes::Node *node : mutable_nodes) {
            if (node->getKind() == nodes::NodeKind::Entrypoint) {
                nodes::NodeEntrypoint *entrypoint = static_cast<nodes::NodeEntrypoint*>(node);
                if (entrypoint->getEntryBytecode() == common::bytecodes::ApicaEntrypointBytecode::Init) {
                    entrypoint_init = entrypoint;
                    break;
                }
            }
        }

        if (entrypoint_init) {
            nodes::NodeCompound *init_compound = static_cast<nodes::NodeCompound*>(entrypoint_init.value()->getBody());
            std::vector<nodes::Node*> &mutable_init_nodes = init_compound->getMutableNodes();
            mutable_init_nodes.insert(mutable_init_nodes.begin(), new nodes::NodeGlobalScope(
                extracted[0]->getPosition(),
                new nodes::NodeCompound(
                    extracted[0]->getPosition(),
                    extracted
                )
            ));
        }
    }
}

void Analyzer::extractSpecifications(nodes::NodeCompound *root) {
    std::vector<nodes::Node*> &mutable_nodes = root->getMutableNodes();
    std::optional<nodes::NodeDataSpecifications*> data_specs = std::nullopt;

    for (auto it = mutable_nodes.begin(); it != mutable_nodes.end(); ++it) {
        nodes::Node* node = *it;

        if (node->getKind() == nodes::NodeKind::DataSpecs) {
            data_specs = static_cast<nodes::NodeDataSpecifications*>(node);
            mutable_nodes.erase(it);
            break;
        }
    }

    if (data_specs) {
        data_specs.value()->addIdCount(utils::IdGenerator::getInstance().getIdCount());
        mutable_nodes.insert(mutable_nodes.begin(), data_specs.value());
    }
}

void Analyzer::setIds(nodes::NodeCompound *root) {
    root->setId();
}